pvlib/irradiance.py

```diff
@@ -2048,3 +2048,70 @@ def _get_dirint_coeffs():
         [0.743440, 0.592190, 0.603060, 0.316930, 0.794390]]
 
     return coeffs[1:, 1:, :, :]
+
+
+def dni(ghi, dhi, zenith, clearsky_dni=None, clearsky_tolerance=1,
+        upper_cutoff_zenith=88.0, lower_cutoff_zenith=80.0, **kwargs):
+    """
+    Determine DNI from GHI and DHI.
+
+    When calculating the DNI from GHI and DHI the calculated DNI may be
+    unreasonably high or negative for zenith angles close to 90 degrees
+    (sunrise/sunset transitions). This function identifies unreasonable DNI
+    values and sets them to NaN. If the clearsky DNI is given unreasonably high
+    values are cut off.
+
+    Parameters
+    ----------
+    ghi : Series
+        Global horizontal irradiance.
+
+    dhi : Series
+        Diffuse horizontal irradiance.
+
+    zenith : Series
+        True (not refraction-corrected) zenith angles in decimal
+        degrees. Angles must be >=0 and <=180.
+
+    clearsky_dni : None or Series
+        Clearsky direct normal irradiance. Default: None.
+
+    clearsky_tolerance : float
+        If 'clearsky_dni' is given this parameter can be used to allow a
+        tolerance by how much the calculated DNI value can be greater than
+        the clearsky value before it is identified as an unreasonable value.
+        Default: 1.
+
+    upper_cutoff_zenith : float
+        Zenith angle above which nonzero DNI values are set to NaN.
+        Default: 88.
+
+    lower_cutoff_zenith : float
+        Zenith angle above which DNI values greater the clearsky DNI (plus
+        allowed tolerance) are corrected. Only applies if 'clearsky_dni' is not
+        None. Default: 80.
+
+    Returns
+    -------
+    dni : Series
+        The modeled direct normal irradiance.
+
+    """
+
+    # calculate DNI
+    dni_tmp = (ghi - dhi) / tools.cosd(zenith)
+    dni = dni_tmp.copy()
+
+    # cutoff negative values
+    dni[dni < 0] = float('nan')
+
+    # set non-zero DNI values for zenith angles >= upper_cutoff_zenith to NaN
+    dni[(zenith >= upper_cutoff_zenith) & (dni != 0)] = float('nan')
+
+    # correct DNI values for zenith angles between lower_cutoff_zenith and
+    # upper_cutoff_zenith that are greater than the clearsky tolerance
+    if clearsky_dni is not None:
+        dni[(zenith >= lower_cutoff_zenith) & (zenith <= upper_cutoff_zenith) &
+            (dni > (clearsky_dni * clearsky_tolerance))] = (clearsky_dni *
```

Comment:
You're doing the same calculation twice, so it may be slightly faster and slightly more readable if you define and use ``max_dni = clearsky_dni * clearsky_tolerance``.